src/sources/kafka.rs

```diff
@@ -0,0 +1,331 @@
+use crate::{
+    event::{self, Event},
+    topology::config::{DataType, GlobalOptions, SourceConfig},
+};
+use bytes::Bytes;
+use futures::{future, sync::mpsc, Future, Poll, Sink, Stream};
+use owning_ref::OwningHandle;
+use rdkafka::{
+    config::ClientConfig,
+    consumer::{Consumer, DefaultConsumerContext, MessageStream, StreamConsumer},
+    error::KafkaResult,
+    message::{BorrowedMessage, Message},
+};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct KafkaSourceConfig {
+    bootstrap_servers: String,
+    topics: Vec<String>,
+    group_id: String,
+    #[serde(default = "default_auto_offset_reset")]
+    auto_offset_reset: String,
+    #[serde(default = "default_session_timeout_ms")]
+    session_timeout_ms: u64,
+    host_key: Option<String>,
+    key_field: Option<String>,
+}
+
+fn default_session_timeout_ms() -> u64 {
+    10000 // default in librdkafka
+}
+
+fn default_auto_offset_reset() -> String {
+    "largest".into() // default in librdkafka
+}
+
+#[typetag::serde(name = "kafka")]
+impl SourceConfig for KafkaSourceConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> Result<super::Source, String> {
+        kafka_source(self.clone(), out)
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+fn kafka_source(
+    config: KafkaSourceConfig,
+    out: mpsc::Sender<Event>,
+) -> Result<super::Source, String> {
+    let consumer = Arc::new(create_consumer(config.clone())?);
+    let source = future::lazy(move || {
+        let hostname = hostname::get_hostname();
+        let host_key = config.host_key.clone().unwrap_or(event::HOST.to_string());
+        let consumer_ref = Arc::clone(&consumer);
+
+        // See https://github.com/fede1024/rust-rdkafka/issues/85#issuecomment-439141656
+        let stream = OwnedConsumerStream {
+            upstream: OwningHandle::new_with_fn(consumer, |c| {
+                let cf = unsafe { &*c };
+                Box::new(cf.start())
+            }),
+        };
+
+        stream
+            .then(move |message| {
+                match message {
+                    Err(e) => Err(error!("Error reading message from Kafka: {:?}", e)),
+                    Ok(Err(e)) => Err(error!("Kafka returned error: {:?}", e)),
+                    Ok(Ok(msg)) => {
+                        let payload = match msg.payload_view::<[u8]>() {
+                            None => return Err(()), // skip messages with empty payload
+                            Some(Err(e)) => return Err(error!("Cannot extract payload: {:?}", e)),
+                            Some(Ok(payload)) => Bytes::from(payload),
+                        };
+
+                        let message_key = if config.key_field.is_some() {
```

Comment:
I'm not sure the best way to consolidate this, but it does feel like there are some duplicate conditionals between here and `create_event`. We check for the presence of `key_field` and the `key_view` in both places.

Reply:
After removing `host_key` option, the `create_event` function became trivial, so I just inlined it.